### DISCUSSION TITLE
ORNO-WE-517 not working

### DISCUSSION THREAD

**Discussion by RichieB2B** [arendst/Tasmota#15957] [Q&A] [answered]
I just tried the https://github.com/laurentppol/Tasmota/blob/development/tasmota/xnrg_17_ornowe517.ino driver and can't get it work. I verified my wiring is correct using the Orno PC software and a CH341SER modbus USB dongle. This works fine.

I compiled Tasmota 12.0.2 with USE_ENERGY_SENSOR (already defined by default) and USE_WE517 defined. I loaded Tasmota on a D1 WEMOS mini and configured TX/GPIO1 as WE517 Tx and RX/GPIO3 as WE517 Rx. I connected TX to the A terminal on the ORNO-WE-517 and RX to the B terminal.

All I get is "error 7" (not enough data). I added some more AddLog statements and it looks like there is never a byte received with a 0x01 value. This is the `WE517_ADDR = mb_address` and should be the start of the relevant modbus frame:

https://github.com/arendst/Tasmota/blob/2fe1d139e3507b4f756928a68b44f9c057a69dae/lib/lib_basic/TasmotaModbus-3.4.0/src/TasmotaModbus.cpp#L89

@woojtekk Did you do anything special to make this work?

```
Jul 10 18:33:14 tasmota-2240 ESP-Project tasmota - Tasmota Version 12.0.2(tasmota)-2_7_4_9(2022-07-10T13:26:38)
Jul 10 18:33:14 tasmota-2240 ESP-ORNO: WE517 We517Modbus->Begin(9600) returned 2
Jul 10 18:33:14 tasmota-2240 ESP-ORNO: WE517 HW serial init 8E1 at 9600 baud
Jul 10 18:33:14 tasmota-2240 ESP-SNS: Hardware Serial
Jul 10 18:33:14 tasmota-2240 ESP-WIF: Attempting connection...
Jul 10 18:33:14 tasmota-2240 ESP-WIF: Connecting to AP1 test Channel 5 BSSId 4A:D9:E7:xx:xx:xx in mode 11n as tasmota-2240...
Jul 10 18:33:14 tasmota-2240 ESP-WIF: Attempting connection...
Jul 10 18:33:14 tasmota-2240 ESP-WIF: Attempting connection...
Jul 10 18:33:14 tasmota-2240 ESP-WIF: Connected
Jul 10 18:33:14 tasmota-2240 ESP-HTP: Web server active on tasmota-2240 with IP address 192.168.3.133
Jul 10 18:33:15 tasmota-2240 ESP-RTC: UTC 2022-07-10T16:33:15, DST 2022-03-27T02:00:00, STD 2022-10-30T03:00:00
Jul 10 18:33:15 tasmota-2240 ESP-RTC: Synced by NTP
Jul 10 18:33:15 tasmota-2240 ESP-MQT: Attempting connection...
Jul 10 18:33:15 tasmota-2240 ESP-MQT: Connected
Jul 10 18:33:15 tasmota-2240 ESP-MQT: tele/tasmota/LWT = Online (retained)
Jul 10 18:33:15 tasmota-2240 ESP-MQT: cmnd/tasmota/POWER = 
Jul 10 18:33:15 tasmota-2240 ESP-MQT: Subscribe to cmnd/tasmota/#
Jul 10 18:33:15 tasmota-2240 ESP-MQT: Subscribe to cmnd/tasmotas/#
Jul 10 18:33:15 tasmota-2240 ESP-MQT: Subscribe to cmnd/DVES_43A8C0_fb/#
Jul 10 18:33:15 tasmota-2240 ESP-MQT: tele/tasmota/INFO1 = {"Info1":{"Module":"Generic","Version":"12.0.2(tasmota)","FallbackTopic":"cmnd/DVES_43A8C0_fb/","GroupTopic":"cmnd/tasmotas/"}}
Jul 10 18:33:15 tasmota-2240 ESP-MQT: tele/tasmota/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-2240","IPAddress":"192.168.3.133"}}
Jul 10 18:33:15 tasmota-2240 ESP-MQT: tele/tasmota/INFO3 = {"Info3":{"RestartReason":"External System","BootCount":96}}
Jul 10 18:33:17 tasmota-2240 ESP-QPC: Reset
Jul 10 18:33:18 tasmota-2240 ESP-NRG: Energy reset by invalid data
Jul 10 18:33:18 tasmota-2240 ESP-APP: Boot Count 96
Jul 10 18:33:19 tasmota-2240 ESP-CFG: Saved to flash at FB, Count 149, Bytes 4096
Jul 10 18:33:20 tasmota-2240 ESP-MQT: tele/tasmota/STATE = {"Time":"2022-07-10T17:33:20","Uptime":"0T00:00:11","UptimeSec":11,"Heap":28,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"MoerasDingen","B
SSId":"4A:D9:E7:91:69:74","Channel":5,"Mode":"11n","RSSI":44,"Signal":-78,"LinkCount":1,"Downtime":"0T00:00:05"}}
Jul 10 18:33:20 tasmota-2240 ESP-MQT: tele/tasmota/SENSOR = {"Time":"2022-07-10T17:33:20","ENERGY":{"TotalStartTime":"2022-03-04T20:44:39","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":[ 0, 0, 0],"Power":[ 0, 0, 0],"ApparentPo
wer":[ 0, 0, 0],"ReactivePower":[ 0, 0, 0],"Factor":[0.00,0.00,0.00],"Voltage":[ 0, 0, 0],"Current":[0.000,0.000,0.000]}}
Jul 10 18:33:22 tasmota-2240 ESP-ORNO: WE517 error 7
Jul 10 18:33:28 tasmota-2240 ESP-ORNO: WE517 error 7
Jul 10 18:33:30 tasmota-2240 ESP-ORNO: WE517 error 7
```

**Comment by RichieB2B** [comment]
I think I'm beginning to understand. I need additional hardware as mentioned in the [Smart Meter Interface](https://tasmota.github.io/docs/Smart-Meter-Interface/) documentation. That was not clear to me before.

>  Additional hardware may be required to read certain measuring devices. For example: RS485toTTL Adaper for Modbus, IR transistor for electricity meters. Sometimes an additional IR diode and resistors.

**Comment by woojtekk** [comment, marked as answer]
Hi, 
I'm sorry for late replay, 
You are right, to communicate with WE-517 you need additional hardware [TTL to RS485 Two-Way Converter Module](https://www.elecrow.com/uart-ttl-to-rs485-twoway-converter-p-1545.html) because  communication is going throw modbus protocol throw RSS485 so you need a converter. 

In my setup this one work for me.  Other setup seems to look fine and should work.   

greetings
Wojtek Z.

**Reply by RichieB2B** [reply]
Thanks for clarifying! This is really poorly documented. The small note I posted above is the only place that mentions it. None of the energy meters at https://github.com/arendst/Tasmota/tree/development/tasmota/tasmota_xnrg_energy have any documentation at https://tasmota.github.io/docs/Supported-Peripherals/

I will order a TTL to RS485 converter and try again.

**Reply by HaraldNoack** [reply]
Have you found a working solution?

If yes, please write your solution in 
https://github.com/arendst/Tasmota/discussions/17545

Thanks!!!!

**Reply by RichieB2B** [reply]
Sorry, I gave up on using Tasmota for this and switched to a Raspberry Pi with https://github.com/volkszaehler/mbmd instead

**Comment by woojtekk** [comment]
hi, 
try [esphome](https://esphome.io). You can read all parameters with easy and simple configuration.
you can also read the electricity consumption recorded and counted by the meter in individual tariffs. In my case, it was a necessary condition.
below link my configuration and to to the list of available parameters.

example to read Voltage L1:
```
sensor:
  - platform: modbus_controller
    modbus_controller_id: epever
    name: "Voltage L1"
    id: VoltL1
    register_type: holding
    address: 0x000E
    unit_of_measurement: "V"
    accuracy_decimals: 2
    value_type: S_QWORD
    lambda: |-
      ESP_LOGD("Modbus Sensor Lambda","Got new data" );
      union {
        float float_value;
        uint32_t raw;
      } raw_to_float;
      if (data.size() < 4 ) {
        ESP_LOGE("Modbus Sensor Lambda", "invalid data size %d",data.size());
        return NAN;
      }
      raw_to_float.raw =   data[0] << 24 | data[1] << 16 | data[2] << 8 |  data[3];
      ESP_LOGD("Modbus Sensor Lambda", "FP32 = 0x%08X => %f", raw_to_float.raw, raw_to_float.float_value);
      return (raw_to_float.float_value);
```

[licznikenergi.yaml.txt](https://github.com/arendst/Tasmota/files/10330557/licznikenergi.yaml.txt)

[OR-WE-517_MODBUS_Registers_List.pdf](https://github.com/arendst/Tasmota/files/10330585/OR-WE-517_MODBUS_Registers_List.pdf)


greetings
Wojtek Z.
